src/main/java/org/eclipse/microprofile/starter/rest/APIService.java

```diff
@@ -0,0 +1,251 @@
+
+    private Response validate(@NotNull Project p) {
+        if (p.getSupportedServer() == null) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .entity(ERROR001)
+                    .type("application/json")
+                    .header("Content-Length", ERROR001.length())
+                    .header("Content-Disposition", "attachment; filename=\"error.json\"")
+                    .build();
+        }
+        if (StringUtils.isBlank(p.getGroupId())) {
+            p.setGroupId(EngineData.DEFAULT_GROUP_ID);
+        }
+        if (StringUtils.isBlank(p.getArtifactId())) {
+            p.setArtifactId(EngineData.DEFAULT_ARTIFACT_ID);
+        }
+        if (p.getMpVersion() == null || p.getMpVersion() == MicroProfileVersion.NONE) {
+            p.setMpVersion(
+                    p.getSupportedServer().getMpVersions().get(p.getSupportedServer().getMpVersions().size() - 1));
+        }
+        if (!mpvToOptions.get(p.getMpVersion()).getSupportedServers().contains(p.getSupportedServer())) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .type("application/json")
+                    .header("Content-Length", ERROR002.length())
+                    .header("Content-Disposition", "attachment; filename=\"error.json\"")
+                    .entity(ERROR002)
+                    .build();
+        }
+        if (p.getJavaSEVersion() == null || p.getJavaSEVersion() == JavaSEVersion.NONE) {
+            p.setJavaSEVersion(EngineData.DEFAULT_JAVA_SE_VERSION);
+        }
+        if (p.getSelectedSpecs() == null || p.getSelectedSpecs().isEmpty()) {
+            p.setSelectedSpecs(mpvToOptions.get(p.getMpVersion()).getSpecs());
+        }
+        if (!mpvToOptions.get(p.getMpVersion()).getSpecs().containsAll(p.getSelectedSpecs())) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                    .type("application/json")
+                    .header("Content-Length", ERROR003.length())
+                    .header("Content-Disposition", "attachment; filename=\"error.json\"")
+                    .entity(ERROR003)
+                    .build();
+        }
+
+        return Response.ok().build();
+    }
+
+    private EngineData getEngineData(Project p) {
+        List<String> selectedSpecs = p.getSelectedSpecs().stream()
+                .map(MicroprofileSpec::getCode).collect(Collectors.toList());
+        EngineData engineData = new EngineData();
+        engineData.getMavenData().setGroupId(p.getGroupId());
+        engineData.getMavenData().setArtifactId(p.getArtifactId());
+        engineData.setTrafficSource(EngineData.TrafficSource.REST);
+        engineData.setSelectedSpecs(selectedSpecs);
+        engineData.setSupportedServer(p.getSupportedServer().getCode());
+        engineData.setMpVersion(p.getMpVersion().getCode());
+        return engineData;
+    }
+
+    private Response processProject(String ifNoneMatch, @NotNull Project p) {
```

Comment:
The @NotNull does nothing here (since it is just called as method) but more important, the @Pattern on groupId and artifactId are not checked/enforced.

Thus requesting a download with `$demo` as artifactId result in a non compilable project.

You can perform programmatic validation by

     Validator validator = Validation.buildDefaultValidatorFactory().getValidator();

The Validator retrieval can be in the init().

    Set<ConstraintViolation<Project>> violations = validator.validate(p);

Validates the Project.

PS. Make sure the - is also allowed in artifactId, see for example org.eclipse.microprofile.starter.validation.PackageValidator

Reply:
Hi, @rdebusscher,

I addressed the validation in a separate commit: 5b56762851914af4062e3116402b775b0d65e0c1 

## Intended changes
* changes validation pattern for artifactId and groupId for Web and REST 
* validation pattern is chnaged to  ```^(?:\w+|\w+[\.-]\w+)+$```  which removes the necessity to swap minuses with dots
* JSF reports the name of the field now
* maximum length is set to 100 for artifactId and the same for groupId
* REST API shares the same validation pattern

## Web changes
Added field name:
![validation1](https://user-images.githubusercontent.com/691097/59074397-81448200-88cb-11e9-8dcd-612f80cf57d4.png)

## Validator, ConstraintViolations
I was looking into adding annotations to the POJO and then using ```validator = Validation.buildDefaultValidatorFactory().getValidator()``` in postConstruct and getting violations in my validate method with ```validator.validate(p)```, but I decided against it as it splits validation to several places.

The current implementation does the basic sanity on deserialization, e.g. when user supplies non-existing Enum. Then all the rest of the validation is done in validate method, including setting default values where permissible.

if you don't like it, I can change it again. if it is the case, I would appreciate some hint on how to do it cleanly. Perhaps ditching validate method and creating validator and some "defaults setter". Not sure.

Reply:
@rdebusscher BTW, I cannot click on that red cross :x: to close the message. Firefox 66.0.5 Linux x86_64. Not sure if this is our fault or a bug for Primefaces...

Reply:
Not a bug, it is just an icon indicating the severity of the message. But I agree, confusing. See #161